physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,590 @@
+	state := FollowerState{
+		AppliedIndex: appliedIndex,
+		LastTerm:     term,
+		NonVoter:     nonVoter,
+	}
+	if appliedIndex > 0 {
+		state.LastHeartbeat = time.Now()
+	}
+
+	s.l.Lock()
+	s.followers[nodeID] = state
+	s.l.Unlock()
+}
+
+// Clear wipes all the information regarding peers in the follower states.
+func (s *FollowerStates) Clear() {
+	s.l.Lock()
+	for i := range s.followers {
+		delete(s.followers, i)
+	}
+	s.l.Unlock()
+}
+
+// Delete the entry of a peer represented by the nodeID from follower states.
+func (s *FollowerStates) Delete(nodeID string) {
+	s.l.Lock()
+	delete(s.followers, nodeID)
+	s.l.Unlock()
+}
+
+// Get retrieves information about a peer from the follower states.
+func (s *FollowerStates) Get(nodeID string) FollowerState {
+	s.l.RLock()
+	state := s.followers[nodeID]
+	s.l.RUnlock()
+	return state
+}
+
+// MinIndex returns the minimum raft index applied in the raft cluster.
+func (s *FollowerStates) MinIndex() uint64 {
+	var min uint64 = math.MaxUint64
+	minFunc := func(a, b uint64) uint64 {
+		if a > b {
+			return b
+		}
+		return a
+	}
+
+	s.l.RLock()
+	for _, state := range s.followers {
+		min = minFunc(min, state.AppliedIndex)
+	}
+	s.l.RUnlock()
+
+	if min == math.MaxUint64 {
+		return 0
+	}
+
+	return min
+}
+
+// Ensure that the Delegate implements the ApplicationIntegration interface
+var _ autopilot.ApplicationIntegration = (*Delegate)(nil)
+
+// Delegate is an implementation of autopilot.ApplicationIntegration interface.
+// This is used by the autopilot library to retrieve information and to have
+// application specific tasks performed.
+type Delegate struct {
+	*RaftBackend
+}
+
+// AutopilotConfig is called by the autopilot library to know the desired
+// autopilot configuration.
+func (d *Delegate) AutopilotConfig() *autopilot.Config {
+	d.l.RLock()
+	config := &autopilot.Config{
+		CleanupDeadServers:      false,
+		LastContactThreshold:    d.autopilotConfig.LastContactThreshold,
+		MaxTrailingLogs:         d.autopilotConfig.MaxTrailingLogs,
+		MinQuorum:               d.autopilotConfig.MinQuorum,
+		ServerStabilizationTime: d.autopilotConfig.ServerStabilizationTime,
+		Ext:                     d.autopilotConfigExt(),
+	}
+	d.l.RUnlock()
+	return config
+}
+
+// NotifyState is called by the autopilot library whenever there is a state
+// change. We update a few metrics when this happens.
+func (d *Delegate) NotifyState(state *autopilot.State) {
+	if d.raft.State() == raft.Leader {
+		metrics.SetGauge([]string{"autopilot", "failure_tolerance"}, float32(state.FailureTolerance))
+		if state.Healthy {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 1)
+		} else {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 0)
+		}
+
+		for id, state := range state.Servers {
+			labels := []metrics.Label{
+				{
+					Name:  "node_id",
+					Value: string(id),
+				},
+			}
+			if state.Health.Healthy {
+				metrics.SetGaugeWithLabels([]string{"autopilot", "healthy"}, 1, labels)
```

Comment:
Sorry to keep coming back to this.  Now there's a problem in that we have autopilot_healthy both with and without labels, which is problematic for Prometheus (maybe others too).  Can we make this one autopilot_healthy_nodes?

Reply:
No problem! Using `autopilot_node_healthy` now.